### PR TITLE
Added error handling to `FunctionGeneratedTableFactory` processing

### LINKED ISSUE #4633: Function-generated tables - unexpected change in generating function crashes server
**Description**

I'm using `function_generated_table` to populate a table from a function that makes API calls. At some point, the API decides I've exceeded its request limit. Therefore, the result of the API call is formatted differently than previous results. Somehow, this crashes my whole server.

**Steps to reproduce**
Run this code exactly:

```
from deephaven import empty_table, function_generated_table

import os
os.system("pip install requests")

import requests
import json

API_KEY = "tHhJLFE1uVKp4VH0YaKlwCNxWFQJFYvc"

def forecast_from_location(location):
    json_response = requests.get(f'https://api.tomorrow.io/v4/weather/forecast?location={location}&apikey={API_KEY}').json()
    minutely_forecast = json_response['timelines']['minutely']
    return(minutely_forecast)

from deephaven import new_table
from deephaven.column import datetime_col, float_col

def table_from_forecast(forecast):
    # unpack json format into list of dicts, then combine dicts into a single dict with list elements, easier to make a table
    unpacked_forecast = [{'time': this_dict['time'], **this_dict['values']} for this_dict in forecast]
    forecast_dict = {k: [dic[k] for dic in unpacked_forecast] for k in unpacked_forecast[0]}
    timestamp = forecast_dict['time']
    del forecast_dict['time']

    table = new_table([
        datetime_col("timestamp", timestamp),
        *[float_col(key, forecast_dict[key]) for key in forecast_dict.keys()]
    ])
    return(table)

def table_from_location(location):
    forecast = forecast_from_location(location)
    return(table_from_forecast(forecast))

LOCATION = "30.2672,-97.7431"
forecast_table = function_generated_table(table_generator = table_from_location,
    refresh_interval_ms = 5*60*1000, kwargs = {'location': LOCATION})

from deephaven import updateby as uby
from deephaven.plot.figure import Figure

cleaned_forecast_table = forecast_table.\
    select(["timestamp", "cloudCover", "humidity", "precipitationProbability",
            "temperature", "temperature", "temperatureApparent", "uvHealthConcern"]).\
    update("temperatureFar = 32 + 9/5 * temperature").\
    update_by(uby.rolling_avg_time("timestamp", "temperatureFarAvg = temperatureFar", "PT00:05:00"))

temp_plot = Figure().plot_xy(series_name="Temperature", t=cleaned_forecast_table, x="timestamp", y="temperatureFar").\
                    plot_xy(series_name="TemperatureAvg", t=cleaned_forecast_table, x="timestamp", y="temperatureFarAvg").\
                    show()
```

I will try to create a smaller minimum reproducible example when I understand the problem better.

**Expected results**

I would expect to see an error in the IDE that says something along the lines of "you've exceeded your API call limit".

**Actual results**

The whole server crashes with the following message:

```
Server shutdown: Exception while processing PeriodicUpdateGraph notification
java.lang.RuntimeException: Error in Python interpreter:
Type: <class 'requests.exceptions.ConnectionError'> Value: HTTPSConnectionPool(host='api.tomorrow.io', port=443):
Max retries exceeded with url: /v4/weather/forecast?location=30.2672,-97.7431&apikey=tHhJLFE1uVKp4VH0YaKlwCNxWFQJFYvc
(Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x2dda54310>:
Failed to resolve 'api.tomorrow.io' ([Errno 8] nodename nor servname provided, or not known)"))
Line: 519 Namespace: send File: /Users/alexpeters/Documents/Core/deephaven-core/core-venv/lib/python3.10/site-packages/requests/adapters.py
Traceback (most recent call last): File "/Users/alexpeters/Documents/Core/deephaven-core/core-venv/lib/python3.10/site-packages/deephaven/table_factory.py", line 401, in table_generator_function
File "<string>", line 33, in table_from_location File "<string>", line 12, in forecast_from_location
File "/Users/alexpeters/Documents/Core/deephaven-core/core-venv/lib/python3.10/site-packages/requests/api.py", line 73, in get
File "/Users/alexpeters/Documents/Core/deephaven-core/core-venv/lib/python3.10/site-packages/requests/api.py", line 59, in request
File "/Users/alexpeters/Documents/Core/deephaven-core/core-venv/lib/python3.10/site-packages/requests/sessions.py", line 589, in request
File "/Users/alexpeters/Documents/Core/deephaven-core/core-venv/lib/python3.10/site-packages/requests/sessions.py", line 703, in send
File "/Users/alexpeters/Documents/Core/deephaven-core/core-venv/lib/python3.10/site-packages/requests/adapters.py", line 519, in send

org.jpy.PyLib.callAndReturnObject(Native Method) org.jpy.PyObject.call(PyObject.java:449)
io.deephaven.integrations.python.JavaLambdaFactory.invoke(JavaLambdaFactory.java:76)
java.base/java.lang.invoke.MethodHandleProxies$1.invoke(MethodHandleProxies.java:193) jdk.proxy1/com.sun.proxy.jdk.proxy1.$Proxy22.get(Unknown Source)
io.deephaven.engine.table.impl.util.FunctionGeneratedTableFactory.updateTable(FunctionGeneratedTableFactory.java:136)
io.deephaven.engine.table.impl.util.FunctionGeneratedTableFactory$FunctionBackedTable.doRefresh(FunctionGeneratedTableFactory.java:190)
io.deephaven.engine.table.impl.util.FunctionGeneratedTableFactory$FunctionBackedTable.run(FunctionGeneratedTableFactory.java:184)
io.deephaven.engine.updategraph.impl.PeriodicUpdateGraph$UpdateSourceRefreshNotification.run(PeriodicUpdateGraph.java:1879)
io.deephaven.engine.updategraph.impl.PeriodicUpdateGraph.runNotification(PeriodicUpdateGraph.java:1322)
io.deephaven.engine.updategraph.impl.PeriodicUpdateGraph$ConcurrentNotificationProcessor.processSatisfiedNotifications(PeriodicUpdateGraph.java:1373)
io.deephaven.engine.updategraph.impl.PeriodicUpdateGraph$NotificationProcessorThreadFactory.lambda$newThread$0(PeriodicUpdateGraph.java:1911)
```

**Versions**
 - Deephaven: Edge
 - OS: MacOS
 - Browser: Safari
 - Docker: Building and running with Gradle

### PR DESCRIPTION
**Pull request by lbooker42**
This change adds error handling to the `FunctionGeneratedTableFactory` call of the user-defined function.  When an exception is detected, we now remove the offending table from the update graph and notify any listeners of the exception vs. the original behavior of delivering a fatal error and shutting down the server.

Closes #4633 

GROOVY CODE USED FOR TESTING/REPRODUCTION
```
import io.deephaven.engine.table.impl.util.FunctionGeneratedTableFactory;
import org.apache.commons.lang3.mutable.MutableInt;

static_table = emptyTable(10).update("X = ii")
MutableInt count = new MutableInt(0);
f2 = () -> {
    if (count.incrementAndGet() == 9) {
        throw new RuntimeException("Deliberate Error");
    }
    return static_table
}

t2 = FunctionGeneratedTableFactory.create(f2, 1000)
```

ORIGINAL ERROR

image

AFTER CODE CHANGE

image

### CODE REVIEW

**Review by alexpeters1208 — APPROVED** [review]
I pulled this branch and ran my code, server did not crash and error is as claimed. LGTM

**Review by rcaudy — DISMISSED** [review]

File: py/server/deephaven/table_factory.py [hunk 1]
```diff
@@ -417,21 +417,23 @@ def table_generator_function():
...
-        if isinstance(source_tables, Table):
-            source_j_tables.append(source_tables.j_table)
-        else:
-            for tbl in source_tables:
-                source_j_tables.append(tbl.j_table)
+        for tbl in source_tables:
+            source_j_tables.append(tbl.j_table)
 
         # Wrap the source_j_tables in a Java array:
         source_j_tables_jarray = jpy.array(_JTable, source_j_tables)
```
Comment (jmao-denver) [inline py/server/deephaven/table_factory.py:431]: Not needed b/c JPY auto matches/converts the Py list to a Java array.

**Review by jmao-denver — COMMENTED** [review]
Not the new change, but there is one place that's not consistent with how vararg is passed the rest of the code base.

**Comment by alexpeters1208** [comment]
> Not the new change, but there is one place that's not consistent with how vararg is handled from the rest of the code base.

Are you referring to the use of `args` and `kwargs` rather than `*args` and `**kwargs`?

File: engine/table/src/main/java/io/deephaven/engine/table/impl/util/FunctionGeneratedTableFactory.java [hunk 2]
```diff
@@ -155,6 +163,15 @@ private FunctionBackedTable getTable() {
 
     private long updateTable() {
         Table newTable = tableGenerator.get();
+        if (newTable.isRefreshing()) {
+            if (ExecutionContext.getContext().getUpdateGraph() != newTable.getUpdateGraph()) {
+                throw new IllegalStateException(
+                        "Function-generated tables must belong to the same UpdateGraph as the creating FunctionGeneratedTableFactory.");
+            }
+            if (!newTable.getUpdateGraph().satisfied(newTable.getUpdateGraph().clock().currentStep())) {
+                throw new IllegalStateException("The function-generated table must be satisfied to be valid.");
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/util/FunctionGeneratedTableFactory.java:172]: `"The function-generated table is not satisfied; did you miss a dependency when specifying source tables?"`

**Review by rcaudy — COMMENTED** [review]
One suggested change.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**